cmd/kube-dns/SPEC.md

```diff
@@ -0,0 +1,140 @@
+    - Answer Format:
+      - `<service>.<ns>.svc.<zone>. <ttl> IN A <cluster-ip>`
+    - Answer Example:
+      - `kubernetes.default.svc.cluster.local. 4 IN A 10.3.0.1`
+  - If a Service named `<service>` exists in `<ns>` without a ClusterIP (i.e., a headless service)  and there are no _ready_ endpoints for that Service, then the result will be empty (i.e., an empty Answer section).
+  - If a Service named `<service>` exists in `<ns>` without a ClusterIP and there are _ready_ endpoints for the Service, the result will be an A record with the specified name and IP for each _ready_ endpoint.
+    - Answer Format:
+      - `<service>.<ns>.svc.<zone>. <ttl> IN A <endpoint-ip>`
+    - Answer Example:
+```
+      headless.default.svc.cluster.local. 4 IN A 10.3.0.1
+      headless.default.svc.cluster.local. 4 IN A 10.3.0.2
+      headless.default.svc.cluster.local. 4 IN A 10.3.0.3
+```
+- **Question:** `<hostname>.<service>.<ns>.svc.<zone>. IN A`
+  - If no Service named `<service>` exists in `<ns>`, the result will be `NXDOMAIN`.
+  - If a Service named `<service>` exists in `<ns>` and a _ready_ endpoint exists for that Service with _hostname_ of `<hostname>` the result will be a single A record with the IP address of that endpoint. If no such endpoint exists, the result will be `NXDOMAIN`.
+    - Answer Format:
+      - `<hostname>.<service>.<ns>.svc.<zone>. <ttl> IN A <endpoint-ip>`
+    - Answer Example:
+      - `my-pet.my-service.default.svc.cluster.local. 4 IN A 10.3.0.100`
+- **Question:** `<a>-<b>-<c>-<d>.<ns>.pod.<zone>. IN A`
+  - If a Pod exists with IP address `<a>.<b>.<c>.<d>`, the result will be an A record with the specified name and the IP address `<a>.<b>.<c>.<d>`.
+    - Answer Format:
+      - `<a>-<b>-<c>-<d>.<ns>.pod.<zone>. <ttl> IN A <a>.<b>.<c>.<d>`
+    - Answer Example:
+      - `10-11-12-13.default.pod.cluster.local. 4 IN A 10.11.12.13`
+  - If no such Pod exists, the result will be `NXDOMAIN`.
+- **Question:** `_<port>._<proto>.<service>.<ns>.svc.<zone>. IN SRV`
+  - If no Service named `<service>` exists in `<ns>`, the result will be `NXDOMAIN`.
+  - If a Service named `<service>` exists in `<ns>` with a ClusterIP, and the service has a port named `<port>` using protocol `<proto>`, the result in the Answer section will be a single SRV record with the port number of the selected port and the name of the A record for the service (`<service>.<ns>.svc.<zone>.`). Unnamed ports will not receive an SRV record. The priority and weight returned are not prescribed by this specification. The Additional section may contain the named A record.
+    - Answer Format:
+       - `_<port>._<proto>.<service>.<ns>.svc.<zone>. <ttl> IN SRV <weight> <priority> <port-number> <service>.<ns>.svc.<zone>.`
+    - Answer Example:
+       - `_https._tcp.kubernetes.default.svc.cluster.local. 30 IN SRV 10 100 443 kubernetes.default.svc.cluster.local.`
```

Comment:
Currently kube-dns sticks an arbitrary identifier in front of the `<service>.<ns>...` name for the A record.